worker/raft/raftclusterer/worker.go

```diff
@@ -0,0 +1,263 @@
+}
+
+func (w *Worker) loop() error {
+	// Get the initial raft cluster configuration.
+	servers, prevIndex, err := w.getConfiguration()
+	if err != nil {
+		return errors.Annotate(err, "getting raft configuration")
+	}
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case details := <-w.serverDetails:
+			prevIndex, err = w.updateConfiguration(servers, prevIndex, details)
+			if err != nil {
+				return errors.Annotate(err, "updating raft configuration")
+			}
+		}
+	}
+}
+
+func (w *Worker) getConfiguration() (map[raft.ServerID]raft.ServerAddress, uint64, error) {
+	future := w.config.Raft.GetConfiguration()
+	prevIndex, err := w.waitIndexFuture(future)
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+	servers := make(map[raft.ServerID]raft.ServerAddress)
+	config := future.Configuration()
+	for _, server := range config.Servers {
+		servers[server.ID] = server.Address
+	}
+	return servers, prevIndex, nil
+}
+
+func (w *Worker) updateConfiguration(
+	configuredServers map[raft.ServerID]raft.ServerAddress,
+	prevIndex uint64,
+	details apiserver.Details,
+) (uint64, error) {
+	newServers := make(map[raft.ServerID]raft.ServerAddress)
+	for _, server := range details.Servers {
+		if server.InternalAddress == "" {
+			continue
+		}
+		serverID := raft.ServerID(names.NewMachineTag(server.ID).String())
+		serverAddress := raft.ServerAddress(server.InternalAddress)
+		newServers[serverID] = serverAddress
+	}
+	if len(newServers) < 3 {
+		// NOTE(axw) when we bootstrap the cluster, we set
+		// the first node's address to be localhost. Until
+		// the cluster is grown beyond a single node, we
+		// can never update the address.
+		//
+		// Changing a node's address can only be done by
+		// removing it from the configuration and adding it
+		// back in with its new address; this cannot be
+		// done if there is only one node, as there must
+		// be one voting node in the configuration at all
+		// times.
+		return prevIndex, nil
+	}
+	ops := w.configOps(configuredServers, newServers)
+	for _, op := range ops {
+		future := op(prevIndex)
+		if err := future.Error(); err != nil {
+			return 0, errors.Trace(err)
+		}
+		prevIndex = future.Index()
+	}
+	return prevIndex, nil
+}
+
+func (w *Worker) configOps(configuredServers, newServers map[raft.ServerID]raft.ServerAddress) []configOp {
+	// This worker can only run on the raft leader, ergo Leader
+	// returns the local address.
+	localAddr := w.config.Raft.Leader()
+
+	// Add new servers first, then remove old servers,
+	// and finish by changing addresses. If the local
+	// server is to be removed, we do that last, to
+	// ensure that the other servers can communicate
+	// amongst themselves.
+	var ops []configOp
+	for id, newAddr := range newServers {
+		if _, ok := configuredServers[id]; ok {
+			continue
+		}
+		logger.Infof("server %q added with address %q", id, newAddr)
+		ops = append(ops, w.addVoterOp(id, newAddr))
+		configuredServers[id] = newAddr
```

Comment:
So... configuredServers is the object being passed in, and we are mutating this as we build up the *proposed* state, before actually doing any calls onto the underlying Raft to make any mutations.
Isn't that racy, where our theoretical state thinks it has been updated, but we end up not applying something.
I guess the idea is that it will trigger an error which will reset everything?

Reply:
> So... configuredServers is the object being passed in, and we are mutating this as we build up the proposed state, before actually doing any calls onto the underlying Raft to make any mutations.
Isn't that racy, where our theoretical state thinks it has been updated, but we end up not applying something.
> I guess the idea is that it will trigger an error which will reset everything?

Right.